provisio-core/src/main/java/ca/vanzyl/provisio/action/artifact/WriteToDiskAction.java

```diff
@@ -41,20 +46,33 @@ public WriteToDiskAction(ProvisioArtifact artifact, File outputDirectory) {
 
     @Override
     public void execute(ProvisioningContext context) {
-        File file = artifact.getFile();
-        if (file != null) {
-            copy(file, new File(outputDirectory, targetArtifactFileName(artifact, file.getName())));
+        if (artifact.getFile() != null) {
+            write(
+                    context,
+                    artifact,
+                    targetArtifactFileName(context, artifact, artifact.getFile().getName()));
         }
     }
 
-    public void copy(File source, File target) {
+    private void write(ProvisioningContext context, ProvisioArtifact source, String targetPath) {
+        File target = new File(outputDirectory, targetPath);
         try {
-            if (!target.getParentFile().exists()) {
-                target.getParentFile().mkdirs();
+            Files.createDirectories(target.getParentFile().toPath());
+            if (ProvisioVariables.allowTargetOverwrite(context)) {
+                if (Files.isRegularFile(target.toPath())) {
+                    logger.warn("Conflict: artifact {} overwrote existing file {}", artifact, targetPath);
+                }
+                Files.copy(artifact.getFile().toPath(), target.toPath(), StandardCopyOption.REPLACE_EXISTING);
+            } else {
+                if (Files.isRegularFile(target.toPath())) {
+                    throw new ProvisioningException(
+                            "Conflict: artifact " + artifact + " would overwrite existing file: " + targetPath);
+                }
+                logger.debug("Copying {} -> {}", artifact, targetPath);
+                Files.copy(artifact.getFile().toPath(), target.toPath());
```

Comment:
can we just not copy if the target file exists and is exactly the same (name, size)?

Reply:
I don't think that's really safe.  The underlying cache stream is quite fast and the big benefit is that it won't write anything if there's no change at all.